### PR TITLE
docs: Provide more information and pointers re. private IP connectivity in new Troubleshooting section in README

### LINKED ISSUE #375: Improve documentation about possible authentication options and required roles, if any?
## Description

I'm trying to understand the _Connection timed out_ below I'm hitting while trying to connect something running as a Service on App Engine Flex using `jdbc:mysql:///DB-NAME?ipTypes=PRIVATE&cloudSqlInstance=PROJECT-NAME:REGION:demo&socketFactory=com.google.cloud.sql.mysql.SocketFactory&user=root&password=...` and am unclear if I need to grant any IAM roles or anything like that for this to work? Or would I need to do some (VPC?) networking magic for the App Engine Flex instance to be able to reach Cloud SQL?

```
java.net.ConnectException: Connection timed out (Connection timed out)
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:289)
	at java.base/sun.security.ssl.BaseSSLSocketImpl.connect(BaseSSLSocketImpl.java:173)
	at com.google.cloud.sql.core.CoreSocketFactory.createSslSocket(CoreSocketFactory.java:275)
	at com.google.cloud.sql.core.CoreSocketFactory.connect(CoreSocketFactory.java:229)
	at com.google.cloud.sql.core.CoreSocketFactory.connect(CoreSocketFactory.java:188)
	at com.google.cloud.sql.mysql.SocketFactory.connect(SocketFactory.java:52)
	at com.google.cloud.sql.mysql.SocketFactory.connect(SocketFactory.java:42)
	at com.mysql.cj.protocol.a.NativeSocketConnection.connect(NativeSocketConnection.java:63)
	at com.mysql.cj.NativeSession.connect(NativeSession.java:144)
	at com.mysql.cj.jdbc.ConnectionImpl.connectOneTryOnly(ConnectionImpl.java:956)
	at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:826)
	at com.mysql.cj.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:456)
	at com.mysql.cj.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:246)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:198)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:200)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:452)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:434)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:65)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Resulted in: com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:151)
	at com.mysql.cj.exceptions.ExceptionFactory.createCommunicationsException(ExceptionFactory.java:167)
	at com.mysql.cj.protocol.a.NativeSocketConnection.connect(NativeSocketConnection.java:89)
	... 14 more
Resulted in: com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
	at com.mysql.cj.jdbc.exceptions.SQLError.createCommunicationsException(SQLError.java:174)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:64)
	at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:836)
	... 11 more
Resulted in: org.jboss.resteasy.spi.UnhandledException: com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:136)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:40)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:97)
	at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

## Solution

> What would you prefer the documentation say? Why would this information be more accurate or helpful? 

Some doc about possible authentication options, e.g. incl. like if a Public IP is a must or not, and how a Private IP on DB can be used.

## Additional Context

FYI Googler here; I've hit this while toying with it for a personal project on the week-end, happy to connect internally.

### PR DESCRIPTION
**Pull request by vorburger**
## Relevant issues:

- Fixes #375

### CODE REVIEW

**Comment by shubha-rajan** [comment]
Hi @vorburger, thanks for the pull request! I agree that a troubleshooting section might be useful for the README, but I'm not sure if including instructions on how to set up serverless VPC access is within the scope of the README for the Java Connector. @kurtisvg, what are your thoughts?

**Comment by vorburger** [comment]
@shubha-rajan I've tried to link to appropriate places, more than provide directly much content here. If you guys strongly feel that it's still a bit excessive as is, it probably could be kept even more succinct, if that's a problem. Let me know if you need anything on particular to be removed.

File: README.md [hunk 1]
```diff
@@ -304,3 +304,13 @@ Example using PostgreSQL:
...
+
+## Networking
+
+`java.net.ConnectException: Connection timed out` failures from (with `The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.` e.g. from the MySQL JDBC drivers) indicate that this Connector could not reach the Cloud SQL database on its public or private IP address.  (See above re. the `ipTypes=PRIVATE` parameter in the JDBC URL.)
+
+If the Cloud SQL database instance is secured and does not have a public but only a private IP to connect to from within a Google Cloud hosted service, then additional configuration in Google Cloud (GCP) may be required to enable private networking connectivity between where the Java code is running and Cloud SQL.  The section about _Connecting from external applications_ in the Cloud SQL Guide for [MySQL](https://cloud.google.com/sql/docs/mysql/connect-overview#external-connection-methods), [PostgreSQL](https://cloud.google.com/sql/docs/postgres/connect-overview#external-connection-methods) and [SQL Server](https://cloud.google.com/sql/docs/sqlserver/connect-overview#external-connection-methods) provides links to detailed documentation about required steps.
```
Comment (kurtisvg) [inline README.md:314]: This is a loaded statement. Having a public IP doesn't make the instance unsecure (and ultimately, having a private IP but not public doesn't make it secure).

**Review by kurtisvg — CHANGES_REQUESTED** [review]

**Comment by kurtisvg** [comment]
> Hi @vorburger, thanks for the pull request! I agree that a troubleshooting section might be useful for the README, but I'm not sure if including instructions on how to set up serverless VPC access is within the scope of the README for the Java Connector. @kurtisvg, what are your thoughts?

I agree. It seems like it would be better to just have a link to the AppEngine docs. It does feel like we should explicitly state that this connector provides authorization and not connectivity (and what that means with effects to public vs private IP access) if we don't already.

**Comment by shubha-rajan** [comment]
Hi @vorburger, we made some updates to the README in #409 that gives more info on private IP connectivity, so I'm going to close this PR for now. Please take a look and let us know if there's any other info you think is important to include. Thanks!

**Pull request closed without merging**